### DISCUSSION TITLE
Repo clones over HTTPS are broken for slow connections

### DISCUSSION THREAD

**Discussion by ell1e** [community/community#135808] [Repositories] [answered]
### Select Topic Area

Question

### Body

**Introduction:** Repo clones over HTTPS are broken for slow connections due to what seems to be an error on GitHub's end, and GitHub apparently doesn't plan to fix it. This is can be a big issue since e.g. this also breaks any submodules or install scripts that indirectly access any GitHub repo of any kind that isn't trivially sized.

I am writing this here so people can reference this issue.

**This is who is affected:**

- Your connection is reliable enough to download 100MB+ files without a hiccup or need to pause and resume in at least the majority of cases.

- You can clone big repositories on gitlab.com and codeberg.org and other alternate git repository platforms without issues, other than at most rare hiccups which don't repeat on 2nd try.

- But your connection is DSL era slow, 5mbit/s or slower.

**This is how it looks like:**

```
$ git clone https://github.com/FortAwesome/Font-Awesome
Cloning into 'Font-Awesome'...
remote: Enumerating objects: 87765, done.
remote: Counting objects: 100% (12255/12255), done.
remote: Compressing objects: 100% (5196/5196), done.
error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 4872 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```

**This is why it seems to be an error on Github's end:**

- I found others with the same problem, although due to people usually having faster internet it's somewhat rare.

- Cloning similarly sized or magnitudes larger repositories with HTTPS from any other platform like gitlab.com works fine, just github.com is affected.

- Cloning the repositories from GitHub via SSH instead works.

- Downloading anything else, like archives, of even larger size via HTTPs from GitHub works. Only HTTPS git clone doesn't.

- **GitHub's support admitted in response to me** that this is likely a problem on their end, but that they don't plan to fix it any time soon. At least this is my understanding of their responses.

**Comment by ell1e** [comment]
This is still broken as of today, 2024-11-07.

**Reply by riv19** [reply]
Yes, not possible to clone repositories on 4G/5G networks. Are these networks outdated ?
And don't try to configure the client side with "http.postBuffer" and similar - doesn't help.

**Reply by ell1e** [reply]
I think it's somehow related to the server having a timeout if in time window X there were less than Y bytes transferred. And this timeout seems to have a really high threshold to clear. This means if the connection isn't like, 20mbit/s or faster, the server will just kill your connection even though it's making progress steadily. At least that's my best guess what could be going on.

**Comment by MarinoJurisic** [comment]
What in the world is this? Some horrible DoS mitigation?

> GitHub's support admitted in response to me that this is likely a problem on their end, but that they don't plan to fix it any time soon

Could I see their response?

This issue can be reliably reproduced by attempting to clone a repository from github.com over HTTPS over Tor. However I have also reproduced this issue by attempting to clone a large repository through connecting to an external SOCKS5 proxy over Tor where the target system hosting the proxy is on a fast network and cloning on the target system itself succeeds which would led credence to your theory.

**Reply by ell1e** [reply]
This is a literal quote from GitHub support:

> I very much agree that the root of this problem appears to be something around how we are implementing HTTPS [...] Unfortunately because we are not getting multiple reports around this issue our networking team is unlikely to make any immediate changes that might help to improve connections experiencing intermittent packet loss like this.

**Reply by MarinoJurisic** [reply]
> Unfortunately because we are not getting multiple reports around this issue our networking team is unlikely to make any immediate changes that might help to improve connections experiencing intermittent packet loss like this.

Think it is worth it for me to submit a second corroborating account of this issue to GitHub support?

**Reply by ell1e** [reply]
I have no idea, but feel free to do so if you got the time. My ticket number was `2835344` if you want to directly reference it.

**Reply by MarinoJurisic** [reply]
Of course they can just ignore it and not change anything. My thinking here is by myself sending in a corroborating account of this issue it removes the 'not getting multiple reports' excuse.

**Reply by MarinoJurisic** [reply]
Think there is a "best" method for contacting GitHub support? Collecting more data (although if they know what they are doing that should not be necessary) and writing a report to send.

**Reply by ell1e** [reply]
I'm not sure, but it's probably [through this form](https://support.github.com/contact?source=subtitle&tags=rr-general-technical). They should already have a network dump so I doubt the issue is missing data, but I could be wrong.

**Reply by MarinoJurisic** [reply]
I have submitted a ticket (`#3107746`) with reference to this discussion and your ticket (`#2835344`).

Now we wait.

**Reply by MarinoJurisic** [reply]
No response from GitHub support regarding my ticket (yet?)

Anyone know the average response time for GitHub support tickets from a free account? (if that makes any difference for priority)

**Reply by airtower-luna** [reply]
A few weeks ago I reported a repository used to host a phishing site (!) via Pages, that took about 3 days to get a response.

**Reply by MarinoJurisic** [reply]
In testing I had a single instance of a large (>1GB) repository clone over Tor complete successfully at ~1MB/s.

I have received a response from GitHub support.

Bit of a curveball here. The issue still occurs and the response reminds me of generic troubleshooting of the symptoms rather than addressing the cause. Considering the issue does not occur with alternate Git hosts over HTTPS it is a non-sequitur (as noted in my original ticket).

I will excerpt their response, both for transparency and for asking if anyone has ideas for potential steps forward.

> Hi MarinoJurisic,
>  
> Thank you for writing in!
>  
> The error message you're seeing, "RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)", is often associated with network issues. Given that you're using Tor or a slow connection, it's likely that the connection is being interrupted or timing out.
> Here are a few things you can try:
> 
>     Increase the Git buffer size: Sometimes, increasing the Git buffer size can help with this issue. You can do this by running the following command in your terminal:
> 
> `git config --global http.postBuffer 524288000`
> 
> This command increases the Git buffer size to 500 MB, which should be more than enough for most repositories.
> 
>     Switch to SSH: If you've previously set up SSH keys, you can use the SSH clone URL instead of HTTPS. This might help if the issue is specifically with HTTPS cloning. For more information, see "About remote repositories."
>     Use SSH over the HTTPS port: If using HTTPS is not an option, and your firewall admin refuses to allow SSH connections, you can try using SSH over the HTTPS port instead.
> 
> If none of these solutions work, it might be a good idea to reach out to your network administrator or ISP to see if they can help troubleshoot the issue.

**Reply by ell1e** [reply]
I sent them a full network dump without TOR or anything, and GitHub support wrote to me that it seems like an issue on their end. I would suggest reminding them of that, perhaps that will help.

**Reply by MarinoJurisic** [reply]
How exactly was the network dump recorded?

I will attempt to press them on the above facts.

**Reply by ell1e** [reply]
I captured according to their instructions from a fresh live usb linux with 100% of the system's network output captured with admin permissions, with a bash script they gave me, and doing a git clone they told me to which showed the problem. I doubt there would be a more complete way to capture it other than physically tapping into the wire or something. Perhaps I'm missing something, but I doubt they lack data to understand how the issue plays out on the client side.

**Reply by ell1e** [reply]
It doesn't seem to contain anything confidential so I hope they won't mind me sharing it here. It's basic instructions for a `tcpdump`. If anybody from GitHub wants me not to share this, please let me know and I'll delete it.

Here are the instructions that I followed at the time:
```
Hey Ellie,
 
To rule out an issue with your local configurations, would you be open to using a live USB of a popular distro to test a clone of an affected repository on your connection?
 
At least it would give us a copy of the environment it occurs in for us to try and reproduce this issue on our end.
 
Cheers,
```
And then: 
```
Hey Ellie,
 
Sure, please run the following script to gather some information and create a support bundle. 
 

#!/bin/bash

# Create a support bundle directory
support_bundle="support_bundle_$(date +%Y%m%d%H%M%S)"
mkdir -p "$support_bundle"

# Gather basic system information
echo "Gathering general system information and time..."
uname -a > "$support_bundle/system_info.txt"
echo "System information gathered."

# Check installed versions of Curl and Git
echo "Checking installed versions of Curl and Git..."
curl --version > "$support_bundle/curl_version.txt"
git --version > "$support_bundle/git_version.txt"
echo "Versions gathered."

# Save Curl and Git configurations
echo "Saving Git configuration.."
git config --list > "$support_bundle/git_config.txt"
echo "Configurations saved."

# Perform git clone operation in verbose mode and record network packets
echo "Cloning repository in verbose mode and recording network packets..."
tcpdump -w "$support_bundle/packets.pcap" &
TCPDUMP_PID=$!
GIT_TRACE=1 GIT_CURL_VERBOSE=1 git clone https://github.com/Gwion/Gwion.git "$support_bundle/repo" > "$support_bundle/git_clone_output.txt" 2>&1
kill $TCPDUMP_PID
echo "Repository cloned in verbose mode and network packets recorded."

# Inform the user of the support bundle creation
echo "Support bundle created at $support_bundle."

 
Any x86/ARM distro with live USB functionality should be fine - we're only using it to exclude any factors that might be due to your personal environment, and so we can try to mimic the conditions on our end to compare what might be different. If you perform any steps prior to running the script please include them, as you'll most likely need to install git/curl as these aren't always included. 
 
The script should create a folder that will include relevant information, and a packet capture, as well as what manages to download of the repository. 
 
Feel free to write back if you have any questions or need a hand.
 
Cheers,
```
No guarantees any of this is safe to do, of course, I'm just forwarding how it came to be. I used an Ubuntu live usb stick at the time, I think. The `tcpdump` is likely fairly privacy invasive, so be careful.

**Reply by MarinoJurisic** [reply]
After replying to GitHub support with reiteration of the above facts, a reference to this discussion and other ticket numbers related to this issue my ticket has been escalated. Hopefully leading to an investigation and resolution.

If anyone from GitHub engineering/support is reading this: Good luck :wave:

**Reply by ell1e** [reply]
Thank you so much. That's amazing to hear.

If you want, let them know I can make new network dumps if required. They can reach me via `el@horse64.org`.

**Reply by MarinoJurisic** [reply]
Just an update. My ticket was filed a month ago. Out of concern over the ticket being forgotten about because of the holiday I sent another message earlier this month and got a personalised response (not a form letter) stating that the engineering team has other issues to attend to but this ticket is in their queue. 

I suppose the next step is for others encountering this issue to file their own tickets and make reference to the previously mentioned tickets and this discussion to ensure it is properly referenced.

**Reply by ell1e** [reply]
I feel like it doesn't matter.  I find concerning that they aren't fixing basic git clones for months, no matter the cause.

**Reply by MarinoJurisic** [reply]
That or something related to load-balancing would be my guess. But ultimately I agree with @ell1e here.

**Reply by MarinoJurisic** [reply]
Sounds like [this other discussion](https://github.com/orgs/community/discussions/149702) is a duplicate of this issue.

**Reply by MarinoJurisic** [reply]
Apparently having a ticket with GitHub Support remain inactive for too long (>3 weeks) sets it to 'archived' which renders it closed. Hopefully just a small bump in the road. 

Created a new ticket with reference to my previous ticket.

**Reply by ell1e** [reply]
The amount of bureaucracy they impose on you for what seems to be something like a load balancer bug is starting to become absurd. :dotted_line_face:

**Reply by MarinoJurisic** [reply]
I was somewhat optimistic about the last response I received but hopefully this will not set me back to square one.

To make matters worse I was not informed of the impending status change of my previous ticket.

**Comment by nikoraasu** [comment]
Same issue here, it's really frustrating

**Reply by ell1e** [reply]
If you [contact GitHub support to report a git clone issue](https://support.github.com/contact?source=subtitle&tags=rr-general-technical) with reference to ticket number `2835344` there's a chance they'll eventually be bothered enough to do something about it. From my limited understanding, they seem to be aware of the issue but under the impression that it's not worth fixing for now.

**Comment by danigar** [comment]
We are experiencing the same issue here since the end of the last week. In our case almost every clone or fetch done through HTTPs is extremely slow (around 1 Mb/s). It's causing all our CI pipelines to fail due timeouts.

In some cases we are seeing the same output error:

```
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```

**Reply by MarinoJurisic** [reply]
All I can suggest is to 

1) create a throwaway GitHub.com account and use SSH for CI repository clones (it seems to be unaffected by this issue) or 

2) Contact GitHub.com support and press them on the facts in this discussion and reference the ticket numbers to (hopefully) escalate the priority of this issue and its resolution.

**Comment by mhaddon** [comment]
We are also having this issue in our CI pipelines hosted on Google Cloud. We are also having extremely slow transfer with SSH. The node itself does not have general networking issues, and the issues are specific to Github.

**Reply by MarinoJurisic** [reply]
In consideration with the previous discussion and your experience it may be prudent to file another ticket with GitHub support making note of ticket numbers and this thread.

**Comment by keyvez** [comment]
Very slow for me too, and continues after weeks, `git clone https://github.com/ltdrdata/ComfyUI-Manager comfyui-manager` Original discussion
: https://github.com/orgs/community/discussions/149702#discussioncomment-12020530 and here is a demo of the slow speed, this happens 8/10 times.

![demo](https://github.com/user-attachments/assets/36801385-d6a6-47fe-887d-33ef2f313dac)

Here are my speed results, so I am not on a slow connection by any means, I have a fiber connection, and ethernet connection to this server, and everything else is blazing fast, even github zip download is fast: 
<img width="403" alt="Screenshot 2025-02-02 at 7 30 42 PM" src="https://github.com/user-attachments/assets/106b2598-e0bf-4755-8869-800ea968408a" />

**Reply by MarinoJurisic** [reply]
The relevant factor here appears to be the throughput of a Git operation over HTTPS transport. Not the speed of your underlying connection.

**Comment by ell1e** [comment, marked as answer]
This seems to have been fixed at some point during the last days leading up to today (2025-03-21), thanks in part to @MarinoJurisic 's tireless efforts to convince Github support to revisit this problem!!! :tada:

**Reply by etbuira** [reply]
I still have the issue
